runtime/src/installed_scheduler_pool.rs

```diff
@@ -0,0 +1,544 @@
+        BankWithSchedulerInner::wait_for_reusable_scheduler(bank, scheduler);
+    }
+
+    pub fn schedule_transaction_executions<'a>(
+        &self,
+        transactions: &[SanitizedTransaction],
+        transaction_indexes: impl Iterator<Item = &'a usize>,
+    ) {
+        trace!(
+            "schedule_transaction_executions(): {} txs",
+            transactions.len()
+        );
+
+        let scheduler_guard = self.inner.scheduler.read().unwrap();
+        let scheduler = scheduler_guard.as_ref().unwrap();
+
+        for (sanitized_transaction, &index) in transactions.iter().zip(transaction_indexes) {
+            scheduler.schedule_execution(&(sanitized_transaction, index));
+        }
+    }
+
+    pub const fn no_scheduler_available() -> InstalledSchedulerRwLock {
+        NO_INSTALLED_SCHEDULER_RW_LOCK
+    }
+
+    #[cfg(any(test, feature = "test-in-workspace"))]
+    pub fn drop_scheduler(&self) {
+        self.inner.drop_scheduler();
+    }
+}
+
+impl BankWithSchedulerInner {
+    #[must_use]
+    fn wait_for_scheduler(&self, reason: WaitReason) -> Option<ResultWithTimings> {
+        Self::do_wait_for_scheduler(&self.bank, &self.scheduler, reason)
+    }
+
+    #[must_use]
+    fn do_wait_for_scheduler(
+        bank: &Bank,
+        scheduler: &InstalledSchedulerRwLock,
+        reason: WaitReason,
+    ) -> Option<ResultWithTimings> {
+        debug!(
+            "wait_for_scheduler(slot: {}, reason: {reason:?}): started...",
+            bank.slot()
+        );
+
+        let mut scheduler = scheduler.write().unwrap();
+        let result_with_timings = if scheduler.is_some() {
+            let result_with_timings = scheduler
+                .as_mut()
+                .and_then(|scheduler| scheduler.wait_for_termination(&reason));
+            if !matches!(reason, WaitReason::ReinitializedForRecentBlockhash) {
+                let scheduler = scheduler.take().expect("scheduler after waiting");
+                scheduler.pool().return_to_pool(scheduler);
+            }
+            result_with_timings
+        } else {
+            None
+        };
+        debug!(
+            "wait_for_scheduler(slot: {}, reason: {reason:?}): finished with: {:?}...",
+            bank.slot(),
+            result_with_timings.as_ref().map(|(result, _)| result)
+        );
+
+        result_with_timings
+    }
+
+    #[must_use]
+    fn wait_for_completed_scheduler_from_drop(&self) -> Option<Result<()>> {
+        let maybe_timings_and_result = self.wait_for_scheduler(WaitReason::DroppedFromBankForks);
+        maybe_timings_and_result.map(|(result, _timings)| result)
+    }
+
+    fn wait_for_reusable_scheduler(bank: &Bank, scheduler: &InstalledSchedulerRwLock) {
+        let maybe_timings_and_result = Self::do_wait_for_scheduler(
+            bank,
+            scheduler,
+            WaitReason::ReinitializedForRecentBlockhash,
+        );
+        assert!(maybe_timings_and_result.is_none());
```

Comment:
`assert!()` outside of tests with no message explaining the constraint being checked is a bit suspicious.
Are you sure this is what you wanted to do?
I would suggest a message with an explanation, specifying which internal invariant must have been violated, should it panic here:
```suggestion
        assert!(
            maybe_timings_and_result.is_none(),
            "Self::do_wait_for_scheduler() must return a result, and it is \
             catastrofic if it did not, because XYZ."
        );
```

Reply:
i agree: 56fc5f5de168e7a37ca0d7e7f49f694afce5cca0